pulsar-broker/src/test/java/org/apache/pulsar/broker/delayed/bucket/BucketDelayedDeliveryTrackerThreadSafetyTest.java

```diff
@@ -0,0 +1,412 @@
+            100, 1000, 100, 10  // Restore original minIndexCountPerBucket for proper testing
+        );
+
+        executorService = Executors.newFixedThreadPool(32);
+    }
+
+    @AfterMethod
+    public void tearDown() throws Exception {
+        if (tracker != null) {
+            tracker.close();
+        }
+        if (executorService != null) {
+            executorService.shutdown();
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Test concurrent containsMessage() calls while adding messages.
+     * This tests the StampedLock optimistic read performance under contention.
+     */
+    @Test
+    public void testConcurrentContainsMessageWithWrites() throws Exception {
+        final int numThreads = 16;
+        final int operationsPerThread = 1000;  // Restore to test bucket creation properly
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch doneLatch = new CountDownLatch(numThreads);
+        final AtomicInteger errors = new AtomicInteger(0);
+        final AtomicReference<Exception> firstException = new AtomicReference<>();
+
+        // Start reader threads
+        for (int i = 0; i < numThreads / 2; i++) {
+            final int threadId = i;
+            executorService.submit(() -> {
+                try {
+                    startLatch.await();
+                    for (int j = 0; j < operationsPerThread; j++) {
+                        long ledgerId = threadId * 1000 + j;
+                        long entryId = j;
+                        // This should not throw exceptions or block indefinitely
+                        tracker.containsMessage(ledgerId, entryId);
+                    }
+                } catch (Exception e) {
+                    errors.incrementAndGet();
+                    firstException.compareAndSet(null, e);
+                    e.printStackTrace();
+                } finally {
+                    doneLatch.countDown();
+                }
+            });
+        }
+
+        // Start writer threads
+        for (int i = numThreads / 2; i < numThreads; i++) {
+            final int threadId = i;
+            executorService.submit(() -> {
+                try {
+                    startLatch.await();
+                    for (int j = 0; j < operationsPerThread; j++) {
+                        long ledgerId = threadId * 1000 + j;
+                        long entryId = j;
+                        long deliverAt = System.currentTimeMillis() + 10000; // 10s delay
+                        tracker.addMessage(ledgerId, entryId, deliverAt);
+                    }
+                } catch (Exception e) {
+                    errors.incrementAndGet();
+                    firstException.compareAndSet(null, e);
+                    e.printStackTrace();
+                } finally {
+                    doneLatch.countDown();
+                }
+            });
+        }
+
+        startLatch.countDown();
```

Comment:
To add more concurrency so that all reader and writer threads proceed to start the work at once, I'd suggest using java.util.concurrency.Phaser instead of the CountDownLatch for coordinating that. There's an example of using Phaser for that purpose here: https://github.com/apache/pulsar/blob/13e0a7b89c64bb516d66a48b0da5ede3cff0d8e2/pulsar-client/src/test/java/org/apache/pulsar/client/impl/MessageIdAdvUtilsTest.java#L55-L72